grouper/repositories/service_account.py

```diff
@@ -0,0 +1,94 @@
+from typing import TYPE_CHECKING
+
+from grouper.entities.group import GroupNotFoundException
+from grouper.entities.service_account import (
+    ServiceAccountNotFoundException,
+    ServiceAccountHasOwnerException,
+)
+from grouper.models.group import Group
+from grouper.models.group_service_accounts import GroupServiceAccount
+from grouper.models.service_account import ServiceAccount as SQLServiceAccount
+from grouper.repositories.interfaces import UserRepository, ServiceAccountRepository
+
+if TYPE_CHECKING:
+    from grouper.graph import GroupGraph
+    from grouper.models.base.session import Session
+
+
+class GraphServiceAccountRepository(ServiceAccountRepository):
+    """Graph-aware storage layer for service accounts."""
+
+    def __init__(self, graph, repository):
+        # type: (GroupGraph, ServiceAccountRepository) -> None
+        self.graph = graph
+        self.repository = repository
+
+    def assign_service_account_to_group(self, username, groupname):
+        # type: (str, str) -> None
+        return self.repository.assign_service_account_to_group(username, groupname)
+
+    def enable_service_account(self, name):
+        # type: (str) -> None
+        return self.repository.enable_service_account(name)
+
+    def set_service_account_description(self, name, description):
+        # type: (str, str) -> None
+        return self.repository.set_service_account_description(name, description)
+
+    def set_service_account_mdbset(self, name, mdbset):
+        # type: (str, str) -> None
+        return self.repository.set_service_account_mdbset(name, mdbset)
+
+
+class SQLServiceAccountRepository(ServiceAccountRepository):
+    """SQL storage layer for service accounts."""
+
+    def __init__(self, session, user_repository):
+        # type: (Session, UserRepository) -> None
+        self.session = session
+        self.user_repository = user_repository
+
+    def assign_service_account_to_group(self, name, groupname):
+        # type: (str, str) -> None
+        service_account = SQLServiceAccount.get(self.session, name=name)
+        if not service_account:
+            raise ServiceAccountNotFoundException(name)
+
+        group = Group.get(self.session, name=groupname)
+        if not group:
+            raise GroupNotFoundException(groupname)
+
+        existing_relationship = GroupServiceAccount.get(
```

Comment:
I'd drop this check and let this function change the ownership. Repositories (and services) should do anything allowed by the schema; checking whether that action is correct is up to the use case.